src/python/WMCore/Database/CMSCouch.py

```diff
@@ -346,6 +347,39 @@ def updateDocument(self, doc_id, design, update_func, fields={}, useBody=False):
                         (self.name, design, update_func, doc_id)
             return self.put(uri=updateUri, data=fields, decode=False)
 
+    def updateBulkDocuments(self, doc_ids, paramsToUpdate, updateLimits=1000):
+
+        uri = '/%s/_bulk_docs/' % self.name
+        conflictDocIDs = []
+        for ids in grouper(doc_ids, updateLimits):
+            # get original documens
+            docs = self.allDocs(options={"include_docs": True}, keys=ids)['rows']
+            data = {}
+            data['docs'] = []
+            for j in docs:
+                doc = {}
+                doc.update(j['doc'])
+                doc.update(paramsToUpdate)
+                data['docs'].append(doc)
+
+            if len(data['docs']) > 0:
+                retval = self.post(uri, data)
+
+                for idx, result in enumerate(retval):
+                    if result.get('error', None) == 'conflict':
+                        conflictDocIDs.append(idx)
+
+        return conflictDocIDs
+
+    def updateBulkDocumentsWithConflictHandle(self, doc_ids, updateParams, updateLimits=1000):
+
+        conflictDocIDs = self.updateBulkDocuments(doc_ids, updateParams, updateLimits)
+        if len(conflictDocIDs) > 0:
+            # waite a second before trying again for the conflit documents
+            time.sleep(1)
```

Comment:
A few comments on this block:
* do you think this sleep is crucial in here? given we fetch the docs with _all_docs, that doesn't require any views/lists and the delay on couch should be closer to 0.
* to avoid an infinite loop, how about we limit it to 5 attempts or so?
* last but not least, how about we log everytime we hit doc conflicts, saying the number of conflicts out of XX docs.

Reply:
I am not sure whether delay is crucial. I think conflict happens when different updates happens at at the same time. Agent doesn't use bulk update yet, but if the too bulk updates happens at the same time it potentially can have a lot of conflicts so I thought we could wait a little to avoid constant update try. 

Other two points are very valid. I will add those.